File: connectors/clients/julia/src/provenance.jl
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
#
# VeriSimDB Julia Client — Provenance operations.
#
# Every octad in VeriSimDB maintains an immutable provenance chain — a
# cryptographically linked sequence of events recording every mutation
# applied to the octad. This file provides functions to query chains,
# record new events, and verify chain integrity.

"""
    get_provenance_chain(client::Client, octad_id::String) -> ProvenanceChain

Retrieve the complete provenance chain for a octad.

The chain is returned in chronological order (oldest event first) and
includes the verification status.

# Arguments
- `client::Client` — The authenticated client.
- `octad_id::String` — The unique identifier of the octad.

# Returns
A `ProvenanceChain` containing all events and verification status.
"""
function get_provenance_chain(client::Client, octad_id::String)::ProvenanceChain
    resp = do_get(client, "/api/v1/octads/$octad_id/provenance")
    return parse_response(ProvenanceChain, resp)
end

"""
    record_provenance(client, octad_id, input) -> ProvenanceEvent

Record a new provenance event on a octad's chain.

The event is cryptographically linked to the previous event in the chain.
The server assigns the event ID and timestamp.

# Arguments
- `client::Client` — The authenticated client.
- `octad_id::String` — The unique identifier of the octad.
- `input::ProvenanceEventInput` — The event details to record.

# Returns
The newly created `ProvenanceEvent` with server-assigned fields.
"""
function record_provenance(
    client::Client,
    octad_id::String,
    input::ProvenanceEventInput
)::ProvenanceEvent
    resp = do_post(client, "/api/v1/octads/$octad_id/provenance", input)
    return parse_response(ProvenanceEvent, resp)
end

"""
    verify_provenance(client::Client, octad_id::String) -> Bool

Verify the cryptographic integrity of a octad's provenance chain.

The server traverses the entire chain, checking each event's hash link to
its parent. Returns `true` if the chain is intact, `false` if tampering
is detected.

# Arguments
- `client::Client` — The authenticated client.
- `octad_id::String` — The unique identifier of the octad.

# Returns
`true` if the provenance chain is verified intact.
"""
function verify_provenance(client::Client, octad_id::String)::Bool
    resp = do_post(client, "/api/v1/octads/$octad_id/provenance/verify", Dict())
    chain = parse_response(ProvenanceChain, resp)
    return chain.verified
end

File: connectors/clients/rust/src/provenance.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>

//! Provenance chain operations.
//!
//! Every hexad entity in VeriSimDB maintains an append-only provenance chain
//! recording creation, transformation, derivation, and access events. This
//! module provides methods to read, append to, and cryptographically verify
//! provenance chains.

use crate::client::VeriSimClient;
use crate::error::Result;
use crate::types::ProvenanceEvent;

impl VeriSimClient {
    /// Retrieve the full provenance chain for a hexad entity.
    ///
    /// Events are returned in chronological order (oldest first).
    ///
    /// # Arguments
    ///
    /// * `id` — The hexad entity identifier.
    ///
    /// # Errors
    ///
    /// Returns [`VeriSimError::NotFound`] if the entity does not exist.
    pub async fn get_provenance_chain(&self, id: &str) -> Result<Vec<ProvenanceEvent>> {
        let path = format!("/api/v1/provenance/{id}");
        self.get(&path).await
    }

    /// Append a new event to a hexad's provenance chain.
    ///
    /// The event is immutably recorded; its timestamp and identifier are
    /// assigned by the server. The returned [`ProvenanceEvent`] contains the
    /// server-assigned fields.
    ///
    /// # Arguments
    ///
    /// * `id`    — The hexad entity identifier.
    /// * `event` — The provenance event to record.
    pub async fn record_provenance(
        &self,
        id: &str,
        event: &ProvenanceEvent,
    ) -> Result<ProvenanceEvent> {
        let path = format!("/api/v1/provenance/{id}");
        self.post(&path, event).await
    }

    /// Verify the integrity of a hexad's provenance chain.
    ///
    /// The server checks that the chain is contiguous, that no events have been
    /// tampered with, and that cryptographic hashes (if enabled) are consistent.
    ///
    /// # Returns
    ///
    /// A JSON object with `valid: bool`, `chain_length: usize`, and optional
    /// `errors` array describing any integrity violations.
    ///
    /// # Arguments
    ///
    /// * `id` — The hexad entity identifier.
    pub async fn verify_provenance(&self, id: &str) -> Result<serde_json::Value> {
        let path = format!("/api/v1/provenance/{id}/verify");
        self.get(&path).await
    }
}
